Documentation/ServicingGranularity.md

```diff
@@ -0,0 +1,146 @@
+decisions like a new version of the .NET Core SDK being released.
+
+> A *product* for the sake of this doc is any released asset where an
+independent decision is made whether that specific asset will be patched and
+shipped.
+
+Each dotnet repo is generally a product, and some repositories contain multiple
+products in a single source tree. This is an example based on 3.0 (some products
+may be missing):
+
+* CoreFX
+  * The NETCoreApp shared framework implementation (product)
+    * `runtime.linux-x64.Microsoft.Private.CoreFx.NETCoreApp`
+    * Always built when NETCoreApp gets any patch.
+  * The out of band (OOB) packages (each is a product)
+    * `System.Data.SqlClient`
+    * `System.Diagnostics.EventLog`
+    * `System.Drawing.Common`
+    * `System.IO.Pipelines`
+    * `System.IO.Ports`
+    * ...more. These `System.*` packages are produced in CoreFX, but aren't
+      included in the NETCoreApp framework. Some are used by downstream
+      frameworks like ASP.NET Core, and by the tools and SDK. They are also used
+      by apps directly via `PackageReference`.
+    * These are somewhat likely to get patches over time.
+
+* Core-Setup
+  * The NETCoreApp shared framework and Runtime/AppHost pack (product)
+    * `Microsoft.NETCore.App.Runtime.linux-x64`
+    * Always built when NETCoreApp gets any patch.
+  * The NETCoreApp Targeting pack (product)
+  * NETStandard Targeting pack (product)
+    * Targeting packs are unlikely to get patches.
+
+* AspNetCore
+  * ASP.NET Core shared framework and Runtime pack (product)
+    * `Microsoft.AspNetCore.App.Runtime.linux-x64`
+  * ASP.NET Core Targeting pack (product)
+
+This means if you assemble a dependency graph of a theoretical .NET Core SDK,
+say 3.0.103, it will contain multiple versions of the same repos:
+
+* .NET Core SDK `v3.0.103`
+  * Core-Setup
+    * `v3.0.3` - shared framework, Runtime/AppHost 
+    * `v3.0.0` - Targeting pack (assuming it hasn't been patched)
+  * CoreFX
+    * `v3.0.3` - System.Security.Permissions, shared framework implementation
+    * `v3.0.2` - System.ServiceProcess.ServiceController
+    * `v3.0.1` - System.Security.Cryptography.Xml
+    * `v3.0.0` - System.IO.Pipelines
+  * [...]
+
+This shows how the number of nodes will naturally grow over time. The max number
+of versions of a specific repo is bounded by the number of products it builds.
+The max is only three for Core-Setup, but larger (dozens?) for CoreFX because of
+the number of independent OOB packages.
+
+## Consequences of depending on multiple versions of the same repo
+
+### Microsoft builds
+This is completely fine for Microsoft servicing builds, because they use
+prebuilts. When building a repo, each upstream product is downloaded from an
+online source, so it doesn't matter which version of the product is being pulled
+down.
+
+### Building from source
+This is bad for source-builds, where downloading prebuilt products is forbidden.
+Every upstream product must be built from source.
```

Comment:
Some (not all) Linux distribution do keep around older distro packages in their distro repository (eg, we build source-build 3.0.10, and 3.0.9 is still around for installation). If that was a general policy (it's not, for example, in Fedora) than source-build could do what Microsoft build does and use older builds. Aside from being not being supported everywhere, it also means source-build needs previous patch versions to have been built in that distro too. A new distro couldn't skip packaging 3.0.0 if 3.0.10 still needs something from there.

Reply:
The new distro onboarding is the issue I have with this... written out in the doc below:

> 1. Say the world has changed in a way that makes it impossible to build anything
>    earlier than CoreFX `v3.0.3` on a new distro or platform. Then, patches need
>    to be applied to `v3.0.0`, `v3.0.1`, and `v3.0.2` to get them to build,
>    because the OOB packages are required. The patching work *probably* only
>    needs to be done once and applied on all affected commits, but as the release
>    branch continues to be serviced, the amount of work to get the product built
>    on a new distro balloons.

(This is also not as big of an issue if we only use this approach for targeting packs, not CoreFX OOB packages.)